Git packaging workflow for Salt (#105)
uyuni-project/uyuni-rfc: accepted/0000-git-salt-packaging.md

- Feature Name: Git-based Salt Packaging
- Start Date: 2025-05-21

# Summary
[summary]: #summary

Package Salt on src.opensuse.org, with one branch per code stream.

# Motivation
[motivation]: #motivation

We want to change our packaging process for Salt, starting with Salt 3008. This is the
first release we package after moving a lot of Salt modules to Salt Extensions. This
approach should meet the following goals:

1. After merging a pull request at [openSUSE/salt](https://github.com/openSUSE/salt), a new package (RPM) is built automatically
2. Building Salt and Salt Extensions can be controlled from a single location (sources may be located elsewhere)
3. We're aligned with the new openSUSE Tumbleweed / SLE 16 workflow packaging
4. We're aligned with upstream's Salt Extension workflows for maintaining, documenting and publish

Package sources for Tumbleweed and SLE 16 will be tracked in git repositories.
[OBS](https://openbuilservice.org) uses the sources from central git forges located at
https://src.opensuse.org and https://src.suse.de respectively. 

## Requirements

We maintain Salt RPMs in different code streams from the same sources. The code streams'

> changelogs differ and release timings can be

Author: ```suggestion
changelogs differ and release timings can be different.
```

- Why are we doing this?
- What use cases does it support?
- What is the expected outcome?

Describe the problem you are trying to solve, and its constraints, without coupling them too closely to the solution you have in mind. If this RFC is not accepted, the motivation can be used to develop alternative solutions.

# Detailed design
[design]: #detailed-design

## `openSUSE/Salt` on Github

 Build metadata (salt.spec, salt.changes, \_multibuild, …) is moved to a subdirectory in
`openSUSE/salt` on GitHub. This allow us to include packaging updates at the time we
create pull requests, e.g. we can include an appropriate changelog together with the
changes.

Files moved to Salt repository:
- pkg/suse/README.SUSE

> - pkg/suse/html.tar.bz2 ???

Author: I checked this and it's still harder than I'd like to build for different OS versions. Since it's static content, I agree with doing it in a single place and keeping the tarball in the repo (or should we use a sub-directory and turn it into a tarball at build time)?

> - pkg/suse/html.tar.bz2 ???

Reviewer: According to what Michael shared in the presentation, it would depend on how frequently we want to update the tar file. In my opinion, we should save it expanded and leverage the new macro to create the tar ball during build time. It would avoid any future issues on potentially using to much disk space on git.

> - pkg/suse/html.tar.bz2 ???

Reviewer 2: As Ricardo said, we are allowed to upload tarballs as long as we are not uploading new tarballs constantly (like daily or weekly).

For this particular case, we should be fine if we upload the tarball once when preparing the 3008.x release, and that would be actually my preferred option here.

The docs tarball contains a ton of html, rst and png files (see the file list [here](https://build.opensuse.org/projects/systemsmanagement:saltstack:products/packages/salt/repositories/16.0/binaries/x86_64/salt-doc-3006.0-lp160.40.1.x86_64.rpm)), and we are not expecting to change the content of the tarball until we switch to a next LTS version of Salt in the future. In that sense, I'd be more in favor of storing the tarball.

> - pkg/suse/html.tar.bz2 ???

Reviewer: Ah thank you Pablo, I didn't notice it was the docs tar. When I wrote it I was thinking about the code, and that would get updated more often i would say.

- pkg/suse/changelogs/sles15sp3.changes
- pkg/suse/changelogs/sles15sp4.changes
- pkg/suse/changelogs/sles15sp5.changes
- pkg/suse/changelogs/sles15sp6.changes
- pkg/suse/changelogs/sles15sp7.changes

### RPM Changelogs

New changelog entries should be part of pull requests. It easy for the code author to
write the user-facing changelog entry while she has all the required context available.

Our changelogs differ between code streams. Most differences are due to different grouping
and entry dates, since we generally keep the package contents in sync.

To help adding new changelog entries in pull requests and update them during rebases, we
add a new Python script `rpmchangelogs`. This script wraps `osc vc` to modify all
`*.changes` files at once. It can `add`, `modify`, and `remove` the latest changelog entry
in all changelogs when the entries are the same.

We use a Github status check to prevent accidental pull requests merges without changelog entries.

## "Package-Git" repository on src.{suse.de,opensuse.org}

Package git: One repository with different branches (see below). The repository contains
openSUSE/salt as a Git submodule.

Branches:
-  `factory` (devel for Tumbleweed)

> -  `products` (why?)

Author: You're right, we need some kind of "products" branch.

-  `testing` (devel for Manager / Uyuni)
-  `sles15sp5` (code stream, src.suse.de)
-  `sles15sp6` (code stream, src.suse.de)
-  `sles15sp7` (code stream, src.suse.de)

> -  `next` (why?)

Reviewer 2: IMO one of the good things of having a `next` environment is that it allows us to put a devel/unstable package there (usually next major version), where we also execute the acceptance CI on it, allowing us to catch possible issues there without touching the `obs://systesmanagement/saltstack/` package.

That way, we keep the `factory` branch and the package at `obs://systesmanagement/saltstack/` being a more stable, where we don't use it for testing unstable-next-versions of salt package.

1. After merging a pull request at [openSUSE/salt](https://github.com/openSUSE/salt), a new package (RPM) is built automatically
2. Building Salt and Salt Extensions can be controlled from a single location (sources may be located elsewhere)
3. We're aligned with the new openSUSE Tumbleweed / SLE 16 workflow packaging
4. We're aligned with upstream's Salt Extension workflows for maintaining, documenting and publish

Package sources for Tumbleweed and SLE 16 will be tracked in git repositories.
[OBS](https://openbuilservice.org) uses the sources from central git forges located at
https://src.opensuse.org and https://src.suse.de respectively. 

## Requirements

We maintain Salt RPMs in different code streams from the same sources. The code streams'
changelogs differ and release timings can be different as well.

# Detailed design
[design]: #detailed-design

## `openSUSE/Salt` on Github

 Build metadata (salt.spec, salt.changes, \_multibuild, …) is moved to a subdirectory in
`openSUSE/salt` on GitHub. This allow us to include packaging updates at the time we
create pull requests, e.g. we can include an appropriate changelog together with the

> changes.

Reviewer 2: I assume this repo will still contain the actual Salt sources, but maybe we can explicity mention it so it is really clear in the picture.

Something like:

```suggestion

We keep maintaining the sources for Salt in this repository on GitHub.

Additionally, the build metadata (salt.spec, salt.changes, \_multibuild, …) is now moved to a subdirectory into this repository. This allow us to include packaging updates at the time we
create pull requests, e.g. we can include an appropriate changelog together with the
changes.
```

- pkg/suse/changelogs/sles15sp4.changes
- pkg/suse/changelogs/sles15sp5.changes
- pkg/suse/changelogs/sles15sp6.changes
- pkg/suse/changelogs/sles15sp7.changes

### RPM Changelogs

New changelog entries should be part of pull requests. It easy for the code author to
write the user-facing changelog entry while she has all the required context available.

Our changelogs differ between code streams. Most differences are due to different grouping
and entry dates, since we generally keep the package contents in sync.

To help adding new changelog entries in pull requests and update them during rebases, we
add a new Python script `rpmchangelogs`. This script wraps `osc vc` to modify all
`*.changes` files at once. It can `add`, `modify`, and `remove` the latest changelog entry
in all changelogs when the entries are the same.

We use a Github status check to prevent accidental pull requests merges without changelog entries.

## "Package-Git" repository on src.{suse.de,opensuse.org}

The canonical package git for distribution code streams will be `pool/salt`. We will have
a devel organisation, with a fork of `pool/salt`: `salt/salt`. Our fork containes the same

> set of branches as `pool/salt` and is used to test and develop Salt together with extensions.

Reviewer 2: Not all our target codestreams are existing as branches in the `pool` org, as for now only those products migrated to the new Git workflow are there. For instance, branches for SLE15SP* are not existing in "pool".

I would suggest something like:

```suggestion
a devel organisation, with a fork of `pool/salt`: `salt/salt`. Our fork containes the same
set of branches as `pool/salt` and additional branches for each codestream that is not yet managed using the new Git workflow (like SLE15SP*). We use this to test and develop Salt together with extensions.
```

If policy allows, openSUSE/salt is a Git submodule, else it's a tarball of the repo at a

> specific revision.

Reviewer 2: I think a submodule won't be really possible because of the sha1-sha256 compatibility.

Additionally, pushing a new tarball to gitea on every new PR is not desired, we should instead just push the entire sources files from GitHub to a `salt` directory inside the `src.opensuse.org/salt/salt` repo, and then use the `#!CreateArchive: %{name}` trick in the spec file. That way, the build service will create the tarball on runtime and we will save storage on gitea.

Author: > I think a submodule won't be really possible because of the sha1-sha256 compatibility.

Yes, it's a pity. At first it sounded like the compatibility code was added on purpose, but apparently that's not the case and will be removed.

> Additionally, pushing a new tarball to gitea on every new PR is not desired, we should instead just push the entire sources files from GitHub to a salt directory [...]

Thanks, I wasn't aware that subdirectories are preferred over tarballs.

### Branches in `src.opensuse.org/pool/salt`
- `factory` (Tumbleweed)
- `leap-16.0`(Leap 16.0)
-  `uyuni` (linked to MLM / Uyuni via scmsync or Git submodule)

Our fork `src.opensuse.org/salt/salt` mirrors `pool/salt`, each branch here is considered

> the devel branch of the corresponding branch in `pool/salt`.

Reviewer 2: I think we should adapt this sentence, as mentioned before, not all our branches are existing in `pool/salt` (only the products already using the new git workflow).

Author: I would prefer to have `leap-16.0` and `uyuni` branches in `src.opensuse.org/pool/salt`.  According to https://src.opensuse.org/pool, `leap-16.0` is an official branch already and I thought we'll migrate Uyuni soon.

Reviewer 2: AFAIK the `leap-16.0` is only used for packages that are part of Leap 16.0 but NOT part of `slfo-1.2`. That is explained here: https://en.opensuse.org/openSUSE:Packaging_for_Leap#Package_sources

For example, the current `salt` package https://src.opensuse.org/pool/salt doesn't contain the `leap-16.0` package, but only `slfo-*` branches.

About a potential `uyuni` branch, that could be, but maybe it is not needed if the uyuni server container is based on Leap and we take the package from the `slfo-*` branch.

### Branches in `src.suse.de/pool/salt`
- `slfo-1.1`(SL Micro 6.1)
- `slfo-1.2`(SL Micro 6.2)
-  `sles15sp3` (SLE 15 SP3)
-  `sles15sp4` (SLE 15 SP4)
-  `sles15sp5` (SLE 15 SP5 + SLE 15 SP6)

> -  `sles15sp7` (SLE 15 SP7)

Reviewer 2: Same consideration here. These branches would only exist in our devel package at `src.opensuse.org/salt/salt` but not in `pool/salt`.

Author: This is a bit different as the SLE 15 code streams are not (yet) maintained in Git and it will take longer. 

For us it would be nice to maintain everything in git, even the SLE 15 code streams. I'll take a look into how we can use it in the transition period.

###  Packaging sources in pool/salt repo

``` text
.gitattributes            # created with obs-git-init
.gitignore                # created with obs-git-init

> .gitmodules               # contains Git submodule status

Reviewer 2: Probably not needed if we are not using submodule.

> salt                      # Git submodule

Reviewer 2: Probably something like:

```suggestion
salt                      # subdirectory with Salt sources from Github
```

README.SUSE               # extracted from `salt` Git submodule
_multibuild               # extracted from `salt` Git submodule
html.tar.bz2              # extracted from `salt` Git submodule
salt-tmpfiles.d           # extracted from `salt` Git submodule
salt.spec                 # extracted from `salt` Git submodule
salt.changes              # extracted from `salt` Git submodule,for the given branch
transactional_update.conf # extracted from `salt` Git submodule

> update-documentation.sh   # extracted from `salt` Git submodule

Reviewer 2: Probably something like: 

```suggestion
README.SUSE               # extracted from `salt` subdirectory
_multibuild               # extracted from `salt` subdirectory
html.tar.bz2              # extracted from `salt` subdirectory
salt-tmpfiles.d           # extracted from `salt` subdirectory
salt.spec                 # extracted from `salt` subdirectory
salt.changes              # extracted from `salt` subdirectory,for the given branch
transactional_update.conf # extracted from `salt` subdirectory
update-documentation.sh   # extracted from `salt` subdirectory
```

```

## "Project-Git" repository on src.{suse.de,opensuse.org}

We use a single "Project-Git" repository, again with one branch per code stream. Packages
in this project are included as Git submodules, checked out at the corresponding branch. 

Per convention, the project git repository is located in the "salt" organisation and
called `_ObsPrj`. An example organisation with the same layout (except that it uses a
singular `master` branch in `_ObsPrj`) is [lua](https://src.opensuse.org/lua)

This project is used to stage changes to Salt together with changes to Salt Extensions. 

### Packages

``` text
salt
salt-ext-zypper
salt-ext-transactional_update
salt-ext-rebootmgr
...
```

## Build Service project on build.{suse.de,opensuse.org}

Build Service projects configure build repositories via it's `meta`. The rest (including
`prjconf`) is maintained in the "Project-Git".

##  Update End-to-end Workflow

When we merge a PR to a release/ branch in openSUSE/salt, a jenkins job updates the
Package-Git repository in our devel organization.

``` text

> 1. update git submodule

Reviewer 2: Or probably something like pull sources from GitHub to "salt" subdirectory.

2. extract files (salt.spec, \_multibuild, …)
3. rename <codestream.changes> to salt.changes
4. commit
5. push 
```

This is implemented with a Makefile, the Jenkins job just calls `make` with
required variables set. The same Makefile also defines targets to similarly update Salt
Extensions. 

[`workflow-direct`](https://src.opensuse.org/adamm/autogits) keeps the Project-Git
up-to-date with changes to the Package-Git repositories.

Submissions to distribution codestreams is done via pull requests from the devel project's

> Package-Git repos to their origins in the pool/ organization.

Reviewer 2: ```suggestion
Submissions to distribution codestreams is done via pull requests from the devel project's
Package-Git repos to their origins in the pool/ organization for the products using the new Git workflows. For codestream not yet migrated to the new git workflow we will keep using MRs.
```

## "Package-Git" repository on src.{suse.de,opensuse.org}

The canonical package git for distribution code streams will be `pool/salt`. We will have

> a devel organisation, with a fork of `pool/salt`: `salt/salt`. Our fork containes the same

Author: > Our fork contains the same set of branches as `pool/salt` and is used to test and develop Salt together with extensions.

That's what I currently have, what would make this clearer to you?

> a devel organisation, with a fork of `pool/salt`: `salt/salt`. Our fork containes the same

Author: I used "our fork" because we can control it. `pool/salt` changes need to pass reviews (e.g. autobuild) since `pool/salt` reflects what's in the distribution. We can use "This fork contains..." if that's clearer, no problem!

> a devel organisation, with a fork of `pool/salt`: `salt/salt`. Our fork containes the same

Reviewer 3: IMO neither `our` nor `this` is clear. To make it clear and resolve this comment, I'd suggest to repeat the fork name:

> We will have a devel organisation, with a fork of `pool/salt`: `salt/salt`. The `salt/salt` fork contains the same ...

We do this correctly later on  - repetition in a tech doc is the right way to remove ambiguity.

(Additional note, `contains` in the RFC has a typo.)
